jbmc/src/java_bytecode/remove_java_new.cpp

```diff
@@ -153,22 +153,52 @@ goto_programt::targett remove_java_newt::lower_java_new_array(
 
   const struct_typet &struct_type = to_struct_type(ns.follow(object_type));
 
-  // Ideally we would have a check for `is_valid_java_array(struct_type)` but
-  // `is_valid_java_array is part of the java_bytecode module and we cannot
-  // introduce such dependencies. We do this simple check instead:
-  PRECONDITION(struct_type.components().size() == 3);
+  PRECONDITION(is_valid_java_array(struct_type));
 
   // Init base class:
   dereference_exprt deref(lhs, object_type);
   auto zero_object = zero_initializer(object_type, location, ns);
   CHECK_RETURN(zero_object.has_value());
-  set_class_identifier(
-    to_struct_expr(*zero_object), ns, to_struct_tag_type(object_type));
+  set_class_identifier(to_struct_expr(*zero_object), ns, object_type);
   dest.insert_before(
     next,
     goto_programt::make_assignment(
       code_assignt(deref, *zero_object), location));
 
+  // If it's a reference array we need to set the dimension and element type
+  // fields (if the given type is void then the caller must set these up itself;
+  // this happens in array[reference].clone(), where the type info must be
+  // copied over from the input array)
+  if(
+    java_array_element_type(object_type) !=
+    java_reference_type(java_void_type()))
+  {
+    const auto underlying_type_and_dimension =
+      java_array_dimension_and_element_type(object_type);
+    if(underlying_type_and_dimension.second >= 1)
```

Comment:
:question: should this be an invariant? if we have an array type, shouldn't the dimension always be `> 0`?

Reply:
We also get 0 for primitive arrays; renamed this `java_*reference*_array_....` accordingly

Reply:
That's surprising... doesn't that mean `assert int[] instance Object[]` fails? Yes: as expected in the JVM (but not expected by me).

Reply:
yep, the primitive arrays are their own thing, subtypes of Object but unrelated to Object[]